scripts/coverage_report_jenkins.py

```diff
@@ -51,6 +118,224 @@ def get_build_numbers(client, job_name):
+    if not ssh.run_command('echo "sonar.host.url={}" > {}'.format(server_url, scanner_conf)):
+        raise Exception('Could write scanner conf, {}s'.format(scanner_conf))
+
+    # Now configure the project
+    #
+    # We have sources in two directories:
+    #
+    #   - /opt/rh/cfme-gemset
+    #   - /var/www/miq/vmdb
+    #
+    # It is very important that we set sonar.sources to a comma delimited
+    # list of these directories but as relative paths, relative to /.   If
+    # we configure them as absolute paths it will only see the files /var/www/miq/vmdb.
+    # Don't know why, it just is that way.
+    #
+    # Hear is an example config:
+    #
+    #   sonar.projectKey=CFME5.9-11
+    #   sonar.projectName=CFME-11
+    #   sonar.projectVersion=5.9.0.17
+    #   sonar.language=ruby
+    #   sonar.sources=opt/rh/cfme-gemset,var/www/miq/vmdb
+    project_conf = 'sonar-project.properties'
+    local_conf = os.path.join(log_path.strpath, project_conf)
+    remote_conf = '/{}'.format(project_conf)
+    config_data = '''
+sonar.projectKey=CFME
+sonar.projectName=CFME
+sonar.projectVersion={version}
+sonar.language=ruby
+sonar.sources=opt/rh/cfme-gemset,var/www/miq/vmdb
+'''.format(
+        version=str(appliance.version).strip())
+
+    # Write the config file locally and then copy to remote.
+    print 'Writing {}'.format(local_conf)
+    f = open(local_conf, 'w')
+    f.write(config_data)
+    f.close()
+    print "Copying {} to appliance as {}".format(local_conf, remote_conf)
+    ssh.put_file(local_conf, remote_conf)
+
+
+def run_sonar_scanner(appliance, scanner_dir, timeout):
+    """Run the sonar scanner
+
+    Run the sonar-scanner.
+
+    Args:
+        appliance:  Appliance object
+        scanner_dir:  Installation directory of the sonar-scanner software.
+        timeout:  timeout in seconds.
+
+    Returns:
+        Nothing
+    """
+    print 'Running sonar scan. This may take a while.'
+    print '   timeout={}'.format(timeout)
+    print '   start_time={}'.format(time.strftime('%T'))
+    ssh = appliance.ssh_client
+    scanner_executable = '{}/bin/sonar-scanner'.format(scanner_dir)
+
+    # It's very important that we run the sonar-scanner from / as this
+    # will allow sonar-scanner to have all CFME ruby source code under
+    # one directory as sonar-scanner expects a project to contain all its
+    # source under one directory.
+    cmd = 'cd /; SONAR_SCANNER_OPTS="-Xmx4096m" {} -X'.format(scanner_executable)
+    result = ssh.run_command(cmd, timeout=timeout)
+    if not result:
+        raise Exception("sonar scan failed!\ncmd: {}\noutput: {}".format(cmd, result))
+    print '   end_time={}'.format(time.strftime('%T'))
+
+
+def sonar_scan(appliance, scanner_url, scanner_dir, server_url, timeout):
+    """Run the sonar scan
+
+    In addition to running the scan, handles the installation of the sonar-scanner software.
+
+    Args:
+        appliance:  Appliance object.
```

Comment:
Consider changing these args up a bit, in the `merge_coverage_data` and `pull_merged_coverage_data` methods you're taking an ssh client instance, and are using a context manager for that instance within main().

here this takes an appliance, and then an ssh client instance is created outside of a context manager in `install_sonar_scanner` and `run_sonar_scanner`. I think you'll be better off just passing the ssh instance with or instead of the appliance object.